tempodb/blocklist/poller.go

```diff
@@ -109,23 +116,41 @@ type Poller struct {
 
 	sharder JobSharder
 	logger  log.Logger
+
+	tenantQueues     *flushqueues.ExclusiveQueues
+	tenantQueuesDone sync.WaitGroup
+	tenantsPolled    map[string]time.Time
+
+	perTenantChan          chan *PerTenant
+	perTenantCompactedChan chan *PerTenantCompacted
+	perTenantErrChan       chan error
+
+	blocklist *List
 }
 
 // NewPoller creates the Poller
-func NewPoller(cfg *PollerConfig, sharder JobSharder, reader backend.Reader, compactor backend.Compactor, writer backend.Writer, logger log.Logger) *Poller {
+func NewPoller(cfg *PollerConfig, sharder JobSharder, reader backend.Reader, compactor backend.Compactor, writer backend.Writer, logger log.Logger, blocklist *List) *Poller {
 	return &Poller{
 		reader:    reader,
 		compactor: compactor,
 		writer:    writer,
 
-		cfg:     cfg,
-		sharder: sharder,
-		logger:  logger,
+		cfg:           cfg,
+		sharder:       sharder,
+		logger:        logger,
+		tenantQueues:  flushqueues.New(int(cfg.TenantPollConcurrency), metricTenantQueueLength),
+		tenantsPolled: make(map[string]time.Time, 1000),
+
+		perTenantChan:          make(chan *PerTenant),
+		perTenantCompactedChan: make(chan *PerTenantCompacted),
+		perTenantErrChan:       make(chan error, 1000),
+
+		blocklist: blocklist,
 	}
 }
 
 // Do does the doing of getting a blocklist
-func (p *Poller) Do(previous *List) (PerTenant, PerTenantCompacted, error) {
+func (p *Poller) Do() (PerTenant, PerTenantCompacted, error) {
```

Comment:
Let's combine `Do()` and `Start()`. Right now `Start()` kicks of N goroutines which watch the tenant queue polling.

If we combine these 2 methods than our channels, go routines, mutex, queues, etc all have a very clean lifecycle. We can remove all of those fields from the struct and instead create and destroy them all in `Do()`.

Reply:
I've pushed a commit for this request.  Let's see what CI says.  What do you think of the change?

Reply:
I don't see a need for the channels. Let's do the straightforwad thing:

```
// for all tenants insert jobs

// for N workers start a goroutine

     // pull job from queue

     // pollBlocklist
     
     // add job to blocklist
```

the blocklist and the queue are concurrency safe so need to worry about that. if you get an error just store it in an atomic err or use a mutex to protect a multierror or whatever makes sense.

let's simplify!

Reply:
The current approach works based on the tests as well as a deployment.  Is there an issue with the channels you are seeing?  I quite like the channelized approach here.

Reply:
It's overcomplicated. This can be done in a much more straightforward way.

Reply:
I don't like the changes required to avoid the channels.  Additionally, I think we should start the workers separately as before, and so the workers remain with the poller through the `Do()` iteration.  The startup + alloc on each iteration doesn't sit well with me

Stylistically, the lock -> append -> unlock in a few places I find noisy and less clean that the original implementation.  The channelized version avoided the locking because collection of the data was done outside of a routine, which I appreciated.

Pragmatically, I'm looking at `Dequeue()` which blocks waiting for an op and I don't see a clean way to avoid attempting to read from the queue without knowing the size of the tenants which feels fragile and prone to race.

Happy to chat.

Reply:
I would like the change to look something like this: 
https://github.com/grafana/tempo/compare/main...joe-elliott:code

It reads cleanly top to bottom, the flow of data is simpler, the lifecycles of the objects are constrained, and it is more in line with the existing structure making it easier to review. It is easier to debug and maintain long term.

> The startup + alloc on each iteration doesn't sit well with me

This is one of the best features of keeping it one function. Limiting the lifecycle of these objects reduces potential for bugs.

> Pragmatically, I'm looking at Dequeue() which blocks waiting for an op

Just make a "DequeueWIthoutBlocking()" call. There's no race b/c you've already inserted all tenants. This priority queue has some features unneeded here, it might make sense to just make a simple local one.

Reply:
Thanks for the review.  I'll close this out and reconsider the approach.  In the mean time, if you'd like to PR your changes for review, feel free.